system/Files/FileCollection.php

```diff
@@ -0,0 +1,396 @@
+<?php
+
+/**
+ * This file is part of the CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Files;
+
+use CodeIgniter\Files\Exceptions\FileException;
+use CodeIgniter\Files\Exceptions\FileNotFoundException;
+use Countable;
+use Generator;
+use InvalidArgumentException;
+use IteratorAggregate;
+use Traversable;
+
+/**
+ * File Collection Class
+ *
+ * Representation for a group of files, with utilities for locating,
+ * filtering, and ordering them.
+ */
+class FileCollection implements Countable, IteratorAggregate
+{
+	/**
+	 * The current list of file paths.
+	 *
+	 * @var string[]
+	 */
+	protected $files = [];
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Resolves a full path and verifies it is an actual directory.
+	 *
+	 * @param string $directory
+	 *
+	 * @return string
+	 *
+	 * @throws FileException
+	 */
+	final protected static function resolveDirectory(string $directory): string
+	{
+		if (! is_dir($directory = set_realpath($directory)))
+		{
+			$caller = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1];
+			throw FileException::forExpectedDirectory($caller['function']);
+		}
+
+		return $directory;
+	}
+
+	/**
+	 * Resolves a full path and verifies it is an actual file.
+	 *
+	 * @param string $file
+	 *
+	 * @return string
+	 *
+	 * @throws FileException
+	 */
+	final protected static function resolveFile(string $file): string
```

Comment:
`final` method doesn't need `protected` as cannot be overriden, I think it can be private:

```suggestion
	final private static function resolveFile(string $file): string
```